test/test_foreach.py

```diff
@@ -341,34 +341,65 @@ def test_int_scalar(self, device, dtype):
+                        with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator,"):
+                            expected = [torch_bin_op(t, scalar) for t in tensors]
 
-                if dtype == torch.bool:
+                        # Test In-place
+                        with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator,"):
+                            foreach_bin_op_(tensors, scalar)
+
+                        with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator,"):
+                            [t.sub_(scalar) for t in tensors]
+                        continue
+
+                    res = foreach_bin_op(tensors, scalar)
+                    expected = [torch_bin_op(t, scalar) for t in tensors]
                     self.assertEqual(res, expected)
 
+                    # Test In-place
                     with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
                         foreach_bin_op_(tensors, scalar)
-                    return
 
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        [t.div_(scalar) for t in tensors]
+
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        [t.mul_(scalar) for t in tensors]
+
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        [t.add_(scalar) for t in tensors]
 
+                    with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator, with a bool"):
+                        [t.sub_(scalar) for t in tensors]
+                    continue
+
+                expected = [torch_bin_op(t, scalar) for t in tensors]
+                res = foreach_bin_op(tensors, scalar)
+
+                # In case of In-place division with integers, we can't change the dtype
                 if foreach_bin_op_ == torch._foreach_div_ and dtype in torch.testing.integral_types() and self.device_type == "cpu":
-                    with self.assertRaisesRegex(RuntimeError,
-                                                "can't be cast to the desired output type"):
-                        foreach_bin_op_(tensors, scalar)
-                    return
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        [t.div_(scalar) for t in tensors]
 
-                # TODO[type promotion]: Fix once type promotion is enabled.
-                if dtype in torch.testing.integral_types() and self.device_type == 'cuda':
-                    self.assertEqual(res, [e.to(dtype) for e in expected])
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        torch._foreach_div_(tensors, scalar)
+                    continue
```

Comment:
Why are we checking for `self.device_type == "cpu"` ? This behavior should happen for both CUDA and CPU